helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -188,14 +196,21 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
 
         if (pendingMessage != null && shouldCleanUpPendingMessage(pendingMessage, currentState,
             currentStateOutput.getEndTime(resourceName, partition, instanceName))) {
-          LogUtil.logInfo(logger, _eventId, String.format(
-              "Adding pending message %s on instance %s to clean up. Msg: %s->%s, current state of resource %s:%s is %s",
-              pendingMessage.getMsgId(), instanceName, pendingMessage.getFromState(),
-              pendingMessage.getToState(), resourceName, partition, currentState));
-          if (!pendingMessagesToCleanUp.containsKey(instanceName)) {
-            pendingMessagesToCleanUp.put(instanceName, new HashMap<String, Message>());
+          logAndAddToCleanUp(messagesToCleanUp, pendingMessage, instanceName, resourceName,
+              partition, currentState, PENDING_MESSAGE);
+        }
+
+        for (Message staleMessage : staleMessages) {
+          if (staleMessage == null) {
+            logger.warn("Should not contain a stale message as null");
```

Comment:
Is it really possible that we have a null element in the return set?

Reply:
Shouldn't be possible. Removed.